File: Solution/0074.Search_a_2D_Matrix/0074.Search_a_2D_Matrix.cpp
//法一
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m=matrix.size();
        int n=matrix[0].size();
        int i=0,j=m*n;
        int mid,r,c;
        while(i<j)
        {
            mid=i+(j-i)/2;
            r=mid/n;
            c=mid%n;
            if(matrix[r][c]==target) return true;
            else if(matrix[r][c]>target) j=mid;
            else i=mid+1;
        }
        return false;
    }
};
//法二
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int top=0,bottom=matrix.size();
        int left=0,right=matrix[0].size();
        int m1,m2,res=-1;
        while(top<bottom)
        {
            m1=top+(bottom-top)/2;
            if(matrix[m1][0]>target) bottom=m1;
            else if(matrix[m1][right-1]<target) top=m1+1;
            else if(matrix[m1][0]<=target&&matrix[m1][right-1]>=target)
            {
                res=m1;
                break;
            }
        }
        if(top==bottom)
            return false;
        while(left<right)
        {
            m2=left+(right-left)/2;
            if(matrix[res][m2]==target) return true;
            else if(matrix[res][m2]<target) left=m2+1;
            else if(matrix[res][m2]>target) right=m2;
         }
        return false;     
    }
};

File: Solution/0074.Search_a_2D_Matrix/0074.Search_a_2D_Matrix.py
#法一
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        n=len(matrix[0])
        i,j=0,len(matrix)*n
        while i<j:
            m=i+(j-i)//2
            r=m//n
            c=m%n
            if matrix[r][c]==target:
                return True
            elif matrix[r][c]>target:
                j=m
            else:
                i=m+1
        return False
      
#法二
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        m,n=len(matrix),len(matrix[0])
        x,y,l,r=0,m,0,n
        while x<y:
            m1=x+(y-x)//2
            if matrix[m1][0]>target:
                y=m1
            elif matrix[m1][n-1]<target:
                x=m1+1
            elif matrix[m1][0]<=target and matrix[m1][n-1]>=target:
                res=m1
                break
        if x==y:
            return False
        while l<r:
            m2=l+(r-l)//2
            if matrix[res][m2]==target:
                return True
            elif matrix[res][m2]>target:
                r=m2
            else:
                l=m2+1
        return False
        
